source/extensions/filters/common/attributes/attributes.cc

```diff
@@ -0,0 +1,551 @@
+#include <atomic>
+#include <chrono>
+#include <iterator>
+#include <map>
+#include <memory>
+#include <string>
+
+#include "absl/strings/str_format.h"
+
+#include "envoy/access_log/access_log.h"
+#include "envoy/buffer/buffer.h"
+#include "envoy/http/filter.h"
+#include "envoy/http/header_map.h"
+#include "envoy/stats/sink.h"
+#include "envoy/upstream/cluster_manager.h"
+#include "envoy/network/filter.h"
+
+#include "id.h"
+#include "source/common/grpc/common.h"
+#include "source/common/http/headers.h"
+#include "source/common/protobuf/utility.h"
+#include "source/common/common/assert.h"
+#include "source/common/common/logger.h"
+#include "source/common/common/fmt.h"
+#include "source/common/common/lock_guard.h"
+#include "source/common/http/header_map_impl.h"
+
+#include "source/extensions/filters/common/attributes/attributes.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace Attributes {
+
+Http::RegisterCustomInlineHeader<Http::CustomInlineHeaderRegistry::Type::RequestHeaders>
+    referer_handle = Http::CustomHeaders::get().Referer;
+
+MapValue_Entry* ValueUtil::getOrInsert(MapValue* m, absl::string_view key) {
+  if (m == nullptr) {
+    return nullptr;
+  }
+  auto entries = m->mutable_entries();
+  for (auto iter = entries->pointer_begin(); iter < entries->pointer_end(); iter++) {
+    auto entry = *iter;
+    if (entry->has_key() && entry->key().has_string_value() && entry->key().string_value() == key) {
+      return entry;
+    }
+  }
+
+  // add a new entry to entries
+  MapValue_Entry* e = m->add_entries();
+  auto k = e->mutable_key();
+  k->set_string_value(std::string(key));
+
+  return e;
+}
+
+Value* ValueUtil::mapValue(MapValue* m) {
+  Value* val = new Value();
+  val->set_allocated_map_value(m);
```

Comment:
If I recall, by doing this protobuf will not make a copy of the MapValue, but it will free it when the message is deallocated. Is that what you want, and more importantly, will people using this library understand that? It seems like something that could cause problems down the road unless the behavior is very clear to people using this library.

Reply:
This was my intention, that this function should take ownership of the `MapValue`. Are you suggesting I should instead use the `mutable_map_value()` methods instead? I'll try that out, I feel like that may obviate the need for this class.

Reply:
Actually I'm guessing that you're saying that its inconsistent in its ownership semantics with the rest of the methods here. Point taken.